File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.js
function solution(board) {
    const N = board.length;
    const cost = Array.from({ length: N }, () =>
        Array.from({ length: N }, () => Array(4).fill(Infinity))
    );

    const queue = [];
    const directions = [
        [0, 1],   // → 0
        [1, 0],   // ↓ 1
        [0, -1],  // ← 2
        [-1, 0],  // ↑ 3
    ];

    // 시작점에서 오른쪽, 아래 방향 먼저 시작
    for (let i = 0; i < 2; i++) {
        const [dy, dx] = directions[i];
        const ny = dy;
        const nx = dx;
        if (board[ny]?.[nx] === 0) {
            cost[ny][nx][i] = 100;
            queue.push([ny, nx, i, 100]);
        }
    }

    while (queue.length) {
        const [y, x, dir, currentCost] = queue.shift();

        for (let i = 0; i < 4; i++) {
            const [dy, dx] = directions[i];
            const ny = y + dy;
            const nx = x + dx;

            if (ny >= 0 && ny < N && nx >= 0 && nx < N && board[ny][nx] === 0) {
                const newCost = currentCost + (dir === i ? 100 : 600);
                if (cost[ny][nx][i] > newCost) {
                    cost[ny][nx][i] = newCost;
                    queue.push([ny, nx, i, newCost]);
                }
            }
        }
    }

    return Math.min(...cost[N - 1][N - 1]);
}

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.*;
class Solution {
    static int N;
    static int[][] cost;
    static int[][] dir={{0,1},{1,0},{-1,0},{0,-1}};
    public int solution(int[][] board) {
        N= board.length-1;
        cost = new int[N+1][N+1];
        for(int i=0; i<=N; i++){
            Arrays.fill(cost[i],Integer.MAX_VALUE);
        }
        dfs(0,0,board,-1,0);
        int answer = 0;
        return cost[N][N];
    }
    static void dfs(int x, int y,int[][] board, int lstDir, int price){
        if(x==N&&y==N){
            cost[y][x] = Math.min(cost[y][x],price);
            return;
        }else{
            for(int i=0; i<4; i++){
                int dy = y+dir[i][0];
                int dx = x+dir[i][1];
                if(dy>=0&&dx>=0&&dy<=N && dx<=N && board[dy][dx]!=1){
                    int newCost = (lstDir==i || lstDir ==-1)?price+100:price+600;
                    if(newCost<=cost[dy][dx]){
                        cost[dy][dx]=newCost;
                        dfs(dx,dy,board,i,newCost);
                    }
                }
            }
        }
    }
}